### PR TITLE
Add Locatable

### LINKED ISSUE #1535: Need a new type of `Movable` that lets you see where it actually is
Name:
- `Locatable`
- ...
Methods:
- `get_setpoint() -> SyncOrAsync[Any]` and `get_position() -> SyncOrAsync[Any]`
- `get_setpoint() -> SyncOrAsync[Any]` and `get_readback() -> SyncOrAsync[Any]`
- ...

### PR DESCRIPTION
**Pull request by coretl**
## Description
Adds Locatable protocol

## Motivation and Context
Fixes #1535 

## How Has This Been Tested?
It hasn't, see comments

### CODE REVIEW

**Comment by coretl** [comment]
I was going to change `mvr` to use this, but like `__read_and_stash_a_motor` is sync, and would need to make it either async, or invent a new message for this. I think I favour a new message, what do you think @tacaswell ?

**Comment by coretl** [comment]
Also, I removed the doc reference to `Movable.position` as this replaces it

File: bluesky/plan_stubs.py
Comment (coretl) [inline bluesky/plan_stubs.py:352]: Should `rd` return `setpoint` or `readback`?

File: bluesky/preprocessors.py
Comment (coretl) [inline bluesky/preprocessors.py:1004]: `hasattr` rather than try/except, is this ok?

File: bluesky/preprocessors.py
Comment (coretl) [inline bluesky/preprocessors.py:1032]: I didn't update the pseudo positioners, they still use position

File: bluesky/protocols.py
Comment (coretl) [inline bluesky/protocols.py:277]: `locate` seemed to fit better with `read` and `set` than `get_location`

File: bluesky/run_engine.py
Comment (coretl) [inline bluesky/run_engine.py:456]: I went with a new message in the end

**Comment by coretl** [comment]
Unit tests failing as Ophyd seems to require a min of python3.6, looks like this is fixed in #1487

File: bluesky/run_engine.py
Comment (coretl) [inline bluesky/run_engine.py:1863]: Actually, having the ability to locate multiple objects concurrently would be useful for load/save. Would this be `Msg('locate', *objs)`? Similarity with `set`, but we don't have `Status` objects, so would have to do the asyncio gather within this function

**Review by danielballan — APPROVED** [review]

**Comment by danielballan** [comment]
I intend to merge this on Wednesday if there are no other comments.

File: bluesky/plan_stubs.py
Comment (ZLLentz) [inline bluesky/plan_stubs.py:345]: I think this should be added to the docstring above along with the other descriptions about how a read key is picked for the return value.

Something like, for example:
```
For devices that implement the Locatable protocol, the location is canonical and is returned without parsing the read keys.
```

File: bluesky/tests/test_examples.py
Comment (ZLLentz) [inline bluesky/tests/test_examples.py:454]: thank you for doing this

File: docs/source/hardware.rst
Comment (ZLLentz) [inline docs/source/hardware.rst:187]: It might make sense to keep the `position` attribute in the documentation in some sort of deprecated sense to help people who stumble upon old devices that have not implemented `locate`.

**Review by ZLLentz — APPROVED** [review]
I like all the decisions made here. `locate` is an unambiguous name for the new message and method and is thematically consistent with the other bluesky verbs. Your implementation that fixes the relative set and reset wrappers is very clean.

File: bluesky/run_engine.py
Comment (ZLLentz) [reply bluesky/run_engine.py:1863]: I think extending this in the same way `set` does for multiple objects makes a lot of sense. I don't fully understand the implications of doing the asyncio gather here vs elsewhere.

Comment (coretl) [reply bluesky/run_engine.py:1863]: The `set()` method returns a `Status` object, so `Msg('set', obj)` only takes one device as an argument, and the combining of `Status` objects is done in the RE. `locate()` can be sync or async, and a `Status` object is not suitable as it is meant for long running tasks that don't have a return value. I think doing the grouping of locate objects in the message and doing the gather in this function is cleaner than pushing it up a level to the main RE loop like `set` does makes more sense here, but I'll defer to @tacaswell on this...

File: docs/source/hardware.rst
Comment (danielballan) [reply docs/source/hardware.rst:187]: Agreed.

**Comment by coretl** [comment]
I have rebased on master, and applied the docs changes.

I also allowed multiple objects to be passed to `Msg('locate')`, and it would call `obj.locate()` on them all concurrently. I would welcome a review specifically of e47d78f2285d82fb0bda5217b1f74e8a63fad370 to see if it's a good idea or not.

**Comment by coretl** [comment]
There are only 2 unresolved conversations left here, I'd welcome comments on them. When they're resolved I think this is ready for merge

File: bluesky/preprocessors.py
Comment (klauer) [inline bluesky/preprocessors.py:983]: Though this was merely copied from the original code, I wonder if there's no better destination for such information. 

It'd be surprising to have this print be a side-effect of listifying a plan generator.

File: bluesky/run_engine.py
Comment (klauer) [reply bluesky/run_engine.py]: `Msg("locate", obj1, obj2, ...)` - a single message with multiple objects seems appropriate, especially with the ability to use `gather()` and perform the requests in parallel.

Looking back at the `read` message (which is closer to `locate` than `set` is, in my opinion), args/kwargs to that message currently get passed to `obj.read(*args, **kwargs)`. However, that usage is *maybe* deprecated according to https://github.com/bluesky/bluesky/blob/e47d78f2285d82fb0bda5217b1f74e8a63fad370/bluesky/utils/__init__.py#L1754-L1763 This makes me think that a change of convention for `locate` is probably a good thing. Of course, @tacaswell may disagree...

File: bluesky/plan_stubs.py
Comment (klauer) [reply bluesky/plan_stubs.py:355]: +1 for `readback`

Comment (danielballan) [reply bluesky/plan_stubs.py:355]: Agreed

File: bluesky/run_engine.py
Comment (danielballan) [reply bluesky/run_engine.py]: I agree.

Comment (coretl) [reply bluesky/run_engine.py]: Are we happy with `Msg('locate', obj) -> obj_location` but `Msg('locate', obj1, obj2) -> [obj1_location, obj2_location]`?

**Review by tacaswell — APPROVED** [review]

File: bluesky/run_engine.py
Comment (tacaswell) [inline bluesky/run_engine.py:1869]: I have a weak preference to always return a list even if it is length 1.  I think it is slightly more verbose if you always just use 1, but much nicer if you want to write generic code that is 0 or more.

File: bluesky/preprocessors.py
Comment (coretl) [reply bluesky/preprocessors.py:983]: Captured in #1552

**Comment by tacaswell** [comment]
Responding to https://github.com/bluesky/bluesky/pull/1536#discussion_r1012352071  I think that passing args/kwargs through to `read` was probably a mistake (and not a feature I think we actually use at NSLS-II).  Once you start to do that then the plan becomes (very silently) dependent on only hardware that will take those inputs.  If you are going to make hardware specfiic plans, better set the state of what ever you were going to pass into read via `bps.mv(..., ...)` before the read than _in_ the read.

**Comment by tacaswell** [comment]
A specific example of why the locate message should always return a list is so that


```python
def some_plan(objs: List[Locatable]):
    objs = ...  # maybe passed to us?
    pos_s = yield from bps.locate(*objs)
    for p in pos_s:
       print(p)
```

It is nice to not have to special case that based on the length of `objs`.

**Comment by tacaswell** [comment]
Sorry for the rapid fire short comments (talking to @coretl and @danielballan on slack).

We may want to take a queue from Matplotlib with `plt.subplots()` here and add a `squeeze=True` keyword to the message to control if we try to drop the outer dimension (https://matplotlib.org/stable/api/figure_api.html#matplotlib.figure.Figure.subplots ).  In the mpl case there can be 2 extra dimensions, but I think it still works here.

That way we can keep the niceity for one-of usage, rhyme with `read` in the one-off case, have the option for the more precise behavior, and have a back-compatible path to extending this to `read` in the future.

**Comment by coretl** [comment]
> Sorry for the rapid fire short comments (talking to @coretl and @danielballan on slack).
> 
> We may want to take a queue from Matplotlib with `plt.subplots()` here and add a `squeeze=True` keyword to the message to control if we try to drop the outer dimension (https://matplotlib.org/stable/api/figure_api.html#matplotlib.figure.Figure.subplots ). In the mpl case there can be 2 extra dimensions, but I think it still works here.
> 
> That way we can keep the niceity for one-of usage, rhyme with `read` in the one-off case, have the option for the more precise behavior, and have a back-compatible path to extending this to `read` in the future.

So are you proposing:
- `Msg(`locate`, obj) -> location`
- `Msg(`locate`, obj1, obj2) -> [location1, location2]`
- `Msg(`locate`, obj, squeeze=False) -> [location]`
- `Msg(`locate`, obj1, obj2, squeeze=False) -> error`

**Comment by tacaswell** [comment]
`Msg("locate", obj1, obj2, squeeze=False) -> [location1, location2]` 

The analogy in numpy.squeeze

```
In [3]: import numpy as np

In [4]: np.ones(5).squeeze()
Out[4]: array([1., 1., 1., 1., 1.])

In [5]: np.ones(5).reshape(1, 5).squeeze()
Out[5]: array([1., 1., 1., 1., 1.])

In [6]: np.ones(5).reshape(5, 1).squeeze()
Out[6]: array([1., 1., 1., 1., 1.])

In [7]: np.ones(6).reshape(2, 3).squeeze()
Out[7]:
array([[1., 1., 1.],
       [1., 1., 1.]])

In [8]: np.ones(1).squeeze()
Out[8]: array(1.)

In [9]: np.ones(1).squeeze().shape
Out[9]: ()
```

so would mean "get rid of any singleton dimensions all the way down to a scalar"

**Comment by coretl** [comment]
Perfect, will make it so...

**Comment by coretl** [comment]
This should be ready to go

**Pull request merged**
